### PR TITLE
KAFKA-16286; Notify listener of latest leader and epoch

### PR DESCRIPTION
**Pull request by jsancio**
KRaft was only notifying listeners of the latest leader and epoch when the replica transition to a new state. This can result in the listener never getting notified if the registration happened after it had become a follower.

This problem doesn't exists for the active leader because the KRaft implementation attempts to notified the listener of the latest leader and epoch when the replica is the active leader.

This issue is fixed by notifying the listeners of the latest leader and epoch after processing the listener registration request.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 1]
```diff
@@ -2958,6 +2958,59 @@ public void testHandleCommitCallbackFiresInCandidateState() throws Exception {
...
+        // Register another listener and verify that it is notified of latest epoch
+        RaftClientTestContext.MockListener secondListener = new RaftClientTestContext.MockListener(
+            OptionalInt.of(localId)
+        );
+        context.client.register(secondListener);
+        context.client.poll();
+
+        LeaderAndEpoch expectedLeaderAndEpoch = new LeaderAndEpoch(OptionalInt.of(otherNodeId), epoch);
+        assertEquals(expectedLeaderAndEpoch, secondListener.currentLeaderAndEpoch());
+    }
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:3019]: Both of these tests fail against trunk.

**Review by cmccabe — APPROVED** [review]
LGTM.

Is there any chance we could fire leader change more than once on a transition? (I guess this might be allowable behavior?)

**Comment by jsancio** [comment]
> Is there any chance we could fire leader change more than once on a transition? (I guess this might be allowable behavior?)

Thanks. I updated the KRaft tests to check that there no duplicate notifications.

**Comment by jsancio** [comment]
All of the `raft:test` tests passed. There seem to be some unrelated failures. Running the tests again.

**Pull request merged**